### ISSUE TITLE
FindAllMarkers returns no differentially expressed genes

### ISSUE THREAD

**Issue by azengan** [satijalab/seurat#1097]
Dear Seurat developers,

I am having an issue with detecting differentially expressed genes using "FindAllMarkers" function. It returns no differentially expressed genes (an empty csv file), no matter which test I am using. When I switch to FindMarker to find only cluster 1 markers, it shows "Error in asMethod(object) : 
  Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105". Could you please help me to fix this issue?

Thanks so much!
Ang

Here is code:

#check the size of the object.
> head(x = comb2@cell.names)

[1] "control_AAACCTGAGAACAACT-1" "control_AAACCTGAGACGCACA-1" "control_AAACCTGAGAGATGAG-1" "control_AAACCTGAGATAGGAG-1" "control_AAACCTGAGCCAGAAC-1"
[6] "control_AAACCTGAGCGAGAAA-1"
> table(comb2@meta.data$orig.ident)

l36178 l36179 l36185 l36186 
 24089  16030  22516  22846 
 
 #Find markers only for cluster 1

> cluster1.markers <- FindMarkers(object = comb2, ident.1 = 0, thresh.use = 0.25, 
+                                 test.use = "roc", only.pos = TRUE,num.cores = 3, do.par=T)

Error in asMethod(object) : 
  Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105

> print(x = head(x = cluster1.markers, n = 5))

Error in head(x = cluster1.markers, n = 5) : 
  object 'cluster1.markers' not found
  
  

> #load("comb2_combined_anzmodifed.Rda")

#Find markers for all clusters

> comb2.markers.roc <- FindAllMarkers(object = comb2, min.pct = 0.3, logfc.threshold = 0.25, test="MAST", only.pos = TRUE, do.print = TRUE)
> write.table(comb2.markers.roc, file="comb2_combined_markersROC.csv",sep=",",quote=F,row.names=F)

> comb2.markers <- FindAllMarkers(object = comb2, logfc.threshold = 0.25, only.pos = TRUE, do.print = TRUE)
> write.table(comb2.markers, file="comb2_combined_markers.csv",sep=",",quote=F,row.names=F)

> sessionInfo()
R version 3.5.0 (2018-04-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS: /n/apps/CentOS7/install/r-3.5.0/lib64/R/lib/libRblas.so
LAPACK: /n/apps/CentOS7/install/r-3.5.0/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] dplyr_0.7.6                 reshape2_1.4.3              Seurat_2.3.4                Matrix_1.2-14               cowplot_0.9.4              
 [6] scater_1.8.4                SingleCellExperiment_1.2.0  SummarizedExperiment_1.10.1 DelayedArray_0.6.6          BiocParallel_1.14.2        
[11] matrixStats_0.54.0          GenomicRanges_1.32.6        GenomeInfoDb_1.16.0         IRanges_2.14.11             S4Vectors_0.18.3           
[16] ggplot2_3.0.0               Biobase_2.40.0              BiocGenerics_0.26.0

**Comment by symbiologist** [comment]
I am having this same issue on a 100K cell dataset on Seurat v2.3. 

I think this may be related to memory - when I downsample my dataset to 20K, it has no problems finding markers.

My linux system has 384 GB memory available to it but when I keep track of R's memory usage, it gets up to around 100 GB and then Seurat returns either "No DE genes" or the "Cholmod "problem too large" error.

Could someone please help us with this? 

Thank you!

David

**Comment by polojacky** [comment]
I just encountered the same problem. With MAST I got no DE genes, but when changed to the default test method (Wilcox), I get some DE genes. So for large datasets, maybe we should try the preferred test method, i.e., Kruskal-Wallis test or Wilcox-test.

**Comment by timoast** [comment]
Hi, can you confirm if the same error occurs using the latest version on the develop branch?

**Comment by cesco-lemon** [comment]
Hi everyone,
I'm having a similar issue, I'm running a big dataset (120k cells) on Seurat v2.4. I've RunPCA, RuntSNE and FindClusters and I'm happy with the subclusters I get.

Then I tried to FindAllMarkers and Seurat only spits out markers for cluster0 (which is the first cluster) and no other marker, but does not give me any error.
object.markers <- FindAllMarkers(object = object, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)

Then I tried to get the markers for each cluster separately:
object.markers <- FindMarkers(object = object, ident.1= 0, ident.2= c(1, 2, 3, 4, 5, 6, 7, 8, 9), only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
object.markers <- FindMarkers(object = object, ident.1= 1, ident.2= c(0, 2, 3, 4, 5, 6, 7, 8, 9), only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
object.markers <- FindMarkers(object = object, ident.1= 2, ident.2= c(1, 0, 3, 4, 5, 6, 7, 8, 9), only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
.....

Doing this I do get the markers for cluster0, which makes me think that my code does work, but then get a similar error:
"Error in asMethod(object) :
Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105".

I wouldn't mind running the FindAllMarkers on a subset of cells using max.cells.per.identity but each cluster has a very different number of cells, should I then use:
object.markers <- FindAllMarkers(object = SubsetData(object=object, max.cells.per.ident=30000), only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)

Would this give me a reliable answer given that some of my clusters are very small (biggest= 22797 cells, smallest=896 cells)? I don't want to lose some of them on the way?
